File: front/convex/messages.ts
import { ConvexError, v } from 'convex/values';
import { mutation, query } from './_generated/server';
import { getIdentity, validateKey } from './utils';

export const getById = query({
  args: { id: v.id('messages') },
  handler: async (ctx, args) => {
    const identity = await getIdentity(ctx);

    const message = await ctx.db.get(args.id);
    if (message?.userId !== identity.tokenIdentifier) return null;

    return message;
  },
});

// api only route
export const apiGetById = query({
  args: { apiKey: v.string(), id: v.id('messages') },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const message = await ctx.db.get(args.id);

    return message == null ? null : message;
  },
});

export const getByThreadId = query({
  args: { threadId: v.id('threads') },
  handler: async (ctx, args) => {
    const identity = await getIdentity(ctx);

    const thread = await ctx.db.get(args.threadId);
    if (thread?.userId !== identity.tokenIdentifier) return null;

    const messages = await ctx.db
      .query('messages')
      .withIndex('by_thread', (q) => q.eq('threadId', thread._id))
      .order('asc')
      .collect();

    return { messages };
  },
});

// api only route
export const apiGetByThreadId = query({
  args: { apiKey: v.string(), threadId: v.id('threads') },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const thread = await ctx.db.get(args.threadId);
    if (thread == null) return null;

    const messages = await ctx.db
      .query('messages')
      .withIndex('by_thread', (q) => q.eq('threadId', thread._id))
      .order('asc')
      .collect();

    return { messages };
  },
});

function createTextPart(text: string) {
  return { type: 'text', text } as const;
}

// api only route
export const apiAppendText = mutation({
  args: { apiKey: v.string(), messageId: v.id('messages'), text: v.string() },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const message = await ctx.db.get(args.messageId);
    if (message == null) return null;

    const parts = message.parts;

    // if no parts, add a text part with args.text
    // if last part is text, append args.text to it
    // if last part is not text, add a new text part with args.text
    const lastPart = parts.length === 0 ? null : parts[parts.length - 1];

    if (lastPart == null) {
      parts.push(createTextPart(args.text));
    } else if (lastPart.type === 'text') {
      lastPart.text += args.text;
    } else {
      parts.push(createTextPart(args.text));
    }

    await ctx.db.patch(message._id, {
      parts,
    });

    return { _id: message._id };
  },
});

// api only route
export const apiAppendReasoning = mutation({
  args: {
    apiKey: v.string(),
    messageId: v.id('messages'),
    reasoning: v.string(),
  },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const message = await ctx.db.get(args.messageId);
    if (message == null) return null;

    if (message.role !== 'assistant') {
      throw new ConvexError('Only assistant messages can have reasoning');
    }

    const reasoning = (message.reasoning ?? '') + args.reasoning;

    await ctx.db.patch(message._id, {
      reasoning,
    });

    return { _id: message._id };
  },
});

// api only route
export const apiAppendAnnotations = mutation({
  args: {
    apiKey: v.string(),
    messageId: v.id('messages'),
    annotations: v.array(
      v.object({
        title: v.string(),
        url: v.string(),
        content: v.string(),
      }),
    ),
  },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const message = await ctx.db.get(args.messageId);
    if (message == null) return null;

    if (message.role !== 'assistant') {
      throw new ConvexError('Only assistant messages can have annotations');
    }

    const existing = message.annotations ?? [];
    const newAnnotations = args.annotations.filter(
      (a) =>
        !existing.some(
          (existingAnnotation) =>
            existingAnnotation.title === a.title &&
            existingAnnotation.url === a.url &&
            existingAnnotation.content === a.content,
        ),
    );

    const annotations = [...existing, ...newAnnotations];

    await ctx.db.patch(message._id, {
      annotations,
    });

    return { _id: message._id };
  },
});

// api only route
export const apiComplete = mutation({
  args: {
    apiKey: v.string(),
    messageId: v.id('messages'),
    model: v.string(),
    modelParams: v.optional(
      v.object({
        reasoningEffort: v.string(),
        includeSearch: v.boolean(),
      }),
    ),
    promptTokenCount: v.number(),
    tokenCount: v.number(),
    durationMs: v.number(),
    tokensPerSecond: v.number(),
    timeToFirstTokenMs: v.number(),
  },
  handler: async (
    ctx,
    {
      apiKey,
      messageId,
      model,
      modelParams,
      promptTokenCount,
      tokenCount,
      durationMs,
      tokensPerSecond,
      timeToFirstTokenMs,
    },
  ) => {
    validateKey(apiKey);

    const message = await ctx.db.get(messageId);
    if (message == null) return null;

    if (message.role === 'assistant') {
      await ctx.db.patch(message._id, {
        status: 'complete',
        model,
        modelParams: modelParams != null ? { ...modelParams } : undefined,
        promptTokenCount,
        tokenCount,
        durationMs,
        tokensPerSecond,
        timeToFirstTokenMs,
      });
    } else {
      throw new ConvexError('Only assistant messages can be completed');
    }

    return { _id: message._id };
  },
});

// api only route
export const apiCancel = mutation({
  args: {
    apiKey: v.string(),
    messageId: v.id('messages'),
  },
  handler: async (ctx, args) => {
    validateKey(args.apiKey);

    const message = await ctx.db.get(args.messageId);
    if (message == null) return null;

    if (message.role === 'assistant') {
      await ctx.db.patch(message._id, {
        status: 'cancelled',
      });
    } else {
      throw new ConvexError('Only assistant messages can be cancelled');
    }

    return { _id: message._id };
  },
});

File: api/src/convex/messages.rs
use std::collections::BTreeMap;

use convex::Value;
use serde::{Deserialize, Serialize};
use specta::Type;

use crate::convex::{ConvexClient, Result, convex_mutation, convex_query};
use crate::convex_serde::to_map;

#[derive(Debug, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum MessageStatus {
  Pending,
  Complete,
  Cancelled,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "type")]
pub enum MessagePart {
  Text { text: String },
  Attachment { id: String },
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Role {
  User,
  Assistant,
  System,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
  #[serde(rename = "_id")]
  pub id: String,
  pub thread_id: String,
  pub status: Option<MessageStatus>,
  pub role: Role,
  pub parts: Vec<MessagePart>,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct MessageIdOnly {
  #[serde(rename = "_id")]
  pub _id: String,
}

pub async fn get_by_id(client: &mut ConvexClient, id: String) -> Result<Option<Message>> {
  const GET_BY_ID: &str = "messages:apiGetById";

  convex_query(
    client,
    GET_BY_ID,
    BTreeMap::from([("id".to_string(), Value::String(id))]),
  )
  .await
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MessagesResponse {
  messages: Vec<Message>,
}

pub async fn get_by_thread_id(client: &mut ConvexClient, thread_id: String) -> Result<Vec<Message>> {
  const GET_BY_THREAD_ID: &str = "messages:apiGetByThreadId";

  // Use the correct parameter name that Convex expects (likely camelCase)
  let response: MessagesResponse = convex_query(
    client,
    GET_BY_THREAD_ID,
    BTreeMap::from([("threadId".to_string(), Value::String(thread_id))]),
  )
  .await?;

  Ok(response.messages)
}
#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ReasoningEffort {
  Low,
  Medium,
  High,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ModelParams {
  pub reasoning_effort: Option<ReasoningEffort>,
  pub include_search: bool,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CompleteMessageArgs {
  pub message_id: String,
  pub model: String,
  pub model_params: Option<ModelParams>,
  pub time_to_first_token_ms: f64,
  pub prompt_token_count: f64,
  pub token_count: f64,
  pub duration_ms: f64,
  pub tokens_per_second: f64,
}

pub async fn complete(client: &mut ConvexClient, args: &CompleteMessageArgs) -> Result<bool> {
  const COMPLETE_MESSAGE: &str = "messages:apiComplete";

  let result = convex_mutation::<Option<MessageIdOnly>>(client, COMPLETE_MESSAGE, to_map(args)?).await?;

  Ok(result.is_some())
}

pub async fn append_text(client: &mut ConvexClient, message_id: String, text: String) -> Result<bool> {
  const APPEND_TEXT: &str = "messages:apiAppendText";

  let args = BTreeMap::from([
    ("messageId".to_string(), Value::String(message_id)),
    ("text".to_string(), Value::String(text)),
  ]);

  let result = convex_mutation::<Option<MessageIdOnly>>(client, APPEND_TEXT, args).await?;

  Ok(result.is_some())
}

pub async fn append_reasoning(client: &mut ConvexClient, message_id: String, reasoning: String) -> Result<bool> {
  const APPEND_REASONING: &str = "messages:apiAppendReasoning";

  let args = BTreeMap::from([
    ("messageId".to_string(), Value::String(message_id)),
    ("reasoning".to_string(), Value::String(reasoning)),
  ]);

  let result = convex_mutation::<Option<MessageIdOnly>>(client, APPEND_REASONING, args).await?;

  Ok(result.is_some())
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnnotationArgs {
  pub message_id: String,
  pub annotations: Vec<Annotation>,
}

#[derive(Clone, Serialize, Type)]
#[serde(rename_all = "camelCase")]
#[specta(rename = "AnnotationResponse")]
pub struct Annotation {
  pub title: String,
  pub url: String,
  pub content: String,
}

pub async fn append_annotations(client: &mut ConvexClient, args: AnnotationArgs) -> Result<bool> {
  const APPEND_ANNOTATIONS: &str = "messages:apiAppendAnnotations";

  let result = convex_mutation::<Option<MessageIdOnly>>(client, APPEND_ANNOTATIONS, to_map(&args)?).await?;

  Ok(result.is_some())
}

pub async fn cancel(client: &mut ConvexClient, message_id: String) -> Result<bool> {
  const CANCEL_MESSAGE: &str = "messages:apiCancel";

  let args = BTreeMap::from([("messageId".to_string(), Value::String(message_id))]);

  let result = convex_mutation::<Option<MessageIdOnly>>(client, CANCEL_MESSAGE, args).await?;

  Ok(result.is_some())
}

pub async fn get_until(client: &mut ConvexClient, thread_id: String, until_id: String) -> Result<Option<Vec<Message>>> {
  const GET_UNTIL: &str = "threads:apiGetMessagesUntil";

  let args = BTreeMap::from([
    ("threadId".to_string(), Value::String(thread_id)),
    ("untilId".to_string(), Value::String(until_id)),
  ]);

  convex_query(client, GET_UNTIL, args).await
}
